wasm_memsafety_rtlib.c

```diff
@@ -1,129 +1,171 @@
-#include <stdlib.h>
+/*
+ * We must store the size of every allocated memory region somewhere, because we
+ * need it to untag that region upon freeing, for which we need the size.
+ *
+ * Our Strategy: For each memory allocation, we allocate additional memory
+ * to store the memory's size just before the user(-requested) memory.
+ * When freeing a memory region, we can then access this embedded size
+ * metadata.
+ * When querying a memory region's size for freeing it, we can directly
+ * access this embedded size.
+ * However, users can demand custom alignment, to which we also have to adhere
+ * when allocating the extra memory for the size. This alignment information
+ * is required to calculate the size of the metadata, which is required to
+ * access the metadata (relative to the user pointer). To solve this issue,
+ * we also have to save the size of the metadata as part of the metadata. We
+ * save this metadata size right before the user pointer, making it easily
+ * accessible in free().
```

Comment:
Let me know ewhat you think of the new strategy. Your version didn't need the extra `metadata_size` because I think you didn't handle alignment yet, which slightly complicates things.